src/peer_loop.rs

```diff
@@ -1544,6 +1559,10 @@ impl PeerLoopHandler {
 
                 Ok(KEEP_CONNECTION_ALIVE)
             }
+            PeerMessage::BootstrapStatus(status) => {
+                self.set_bootstrap_status(status).await;
+                Ok(KEEP_CONNECTION_ALIVE)
+            }
```

Comment:
This status should only be set *once* per connection. So I would first check if it is set for the peer in question, and only then acquire a write-lock. If the status is already set, punish peer. We should be very stingy about write locks.

Reply:
I want to give nodes a chance to change their bootstrap status.

Once per connection is probably fine – after all, you need to restart a node to change its bootstrap status – but the way you suggest to go about that is more restrictive than once per connection. Your suggestion means that a node can never update its own bootstrap status, it can only ever set it once. Only when the peer is restarted, and forgets the bootstrap status of all its past peers, can the status be set again.